File: bench-core/node/schema.mjs
// The one versioned result record (result/1), used by JS and (mirrored) by Python. Promotes
// run.mjs's rich record, adds provenance + camera state + gate verdicts, and unifies the
// three legacy shapes. makeRecord assembles from the parts the runner collected and derives
// ok + gateFailures from the gate results; validate() rejects a record missing a gate
// verdict or camera state. A failing gate quarantines (ok:false) — it does NOT drop the row.
export const RESULT_SCHEMA_VERSION = 'result/1';
export const REQUIRED_GATES = ['gpu', 'coverage', 'presentation', 'cameraState', 'inputHonesty'];

const passed = (gr) => !!gr && (gr.pass || gr.verdict === 'n/a');
const verdictStr = (gr) => {
  if (!gr) return 'missing';
  if (gr.verdict === 'n/a') return 'n/a' + (gr.detail ? ': ' + gr.detail : '');
  return (gr.pass ? 'pass' : 'fail') + (gr.detail ? ': ' + gr.detail : '');
};

// parts: { domain, tool, engine, dataset, primitives, scale, coverage, protocol, metrics,
//          presentation, gpu, timings, gateResults: {gpu,coverage,presentation,cameraState,
//          inputHonesty}, provenance, error }
export function makeRecord(parts) {
  const { gateResults = {}, error } = parts;
  const gates = {};
  for (const k of Object.keys(gateResults)) gates[k] = verdictStr(gateResults[k]);
  const ok = !error && Object.values(gateResults).every(passed);
  const gateFailures = Object.entries(gateResults)
    .filter(([, gr]) => !passed(gr))
    .map(([k, gr]) => `${k}: ${gr?.detail || gr?.verdict || 'fail'}`);
  return {
    schemaVersion: RESULT_SCHEMA_VERSION,
    domain: parts.domain,
    tool: parts.tool,
    engine: parts.engine,
    dataset: parts.dataset,
    primitives: parts.primitives,
    scale: parts.scale,
    coverage: parts.coverage,
    protocol: parts.protocol,
    metrics: parts.metrics,
    presentation: parts.presentation,
    gpu: parts.gpu,
    gates,
    ok,
    gateFailures,
    ...(error ? { error: String(error) } : {}),
    provenance: parts.provenance,
    ...(parts.timings ? { timings: parts.timings } : {}),
  };
}

export function validate(rec) {
  const problems = [];
  if (rec.schemaVersion !== RESULT_SCHEMA_VERSION) problems.push(`schemaVersion=${rec.schemaVersion}`);
  if (!rec.protocol?.cameraState) problems.push('protocol.cameraState missing');
  if (!rec.gates) problems.push('gates missing');
  else for (const g of REQUIRED_GATES) if (rec.gates[g] == null) problems.push(`gates.${g} missing`);
  return { valid: problems.length === 0, problems };
}

File: bench-core/py/schema.py
"""result/1 record — Python mirror of bench-core/node/schema.mjs. Same shape so report-gen
consumes JS and native records identically. A failing gate quarantines (ok:false), not drops."""
RESULT_SCHEMA_VERSION = "result/1"
REQUIRED_GATES = ["gpu", "coverage", "presentation", "cameraState", "inputHonesty"]


def _passed(gr):
    return bool(gr) and (gr.get("pass") or gr.get("verdict") == "n/a")


def _verdict_str(gr):
    if not gr:
        return "missing"
    if gr.get("verdict") == "n/a":
        return "n/a" + (": " + gr["detail"] if gr.get("detail") else "")
    return ("pass" if gr.get("pass") else "fail") + (": " + gr["detail"] if gr.get("detail") else "")


def make_record(*, domain, tool, engine, dataset, primitives, scale, coverage, protocol,
                metrics, presentation, gpu, gate_results=None, provenance=None, error=None, timings=None):
    gate_results = gate_results or {}
    gates = {k: _verdict_str(v) for k, v in gate_results.items()}
    ok = (error is None) and all(_passed(v) for v in gate_results.values())
    gate_failures = [
        f"{k}: {(v or {}).get('detail') or (v or {}).get('verdict') or 'fail'}"
        for k, v in gate_results.items() if not _passed(v)
    ]
    rec = {
        "schemaVersion": RESULT_SCHEMA_VERSION, "domain": domain, "tool": tool, "engine": engine,
        "dataset": dataset, "primitives": primitives, "scale": scale, "coverage": coverage,
        "protocol": protocol, "metrics": metrics, "presentation": presentation, "gpu": gpu,
        "gates": gates, "ok": ok, "gateFailures": gate_failures, "provenance": provenance,
    }
    if error is not None:
        rec["error"] = str(error)
    if timings is not None:
        rec["timings"] = timings
    return rec


def validate(rec):
    p = []
    if rec.get("schemaVersion") != RESULT_SCHEMA_VERSION:
        p.append(f"schemaVersion={rec.get('schemaVersion')}")
    if not (rec.get("protocol") or {}).get("cameraState"):
        p.append("protocol.cameraState")
    g = rec.get("gates") or {}
    for k in REQUIRED_GATES:
        if g.get(k) is None:
            p.append(f"gates.{k}")
    return {"valid": len(p) == 0, "problems": p}
